docs(design): operator-prepaid attestation storage (#4011)
near/mpc: docs/design/operator-prepaid-attestation-storage.md

```text
entries an account holds  ≤  grants it has bought
```

Because grants are denominated in *entries*, re-pricing after a storage-price change affects only future grants and can never strand a sold one — the failure mode a NEAR-denominated balance would have.

The counter is **available** grants, not lifetime total, so `0` means either "never prepaid" or "prepaid, now backing an entry"; disambiguate with `get_tee_accounts`. Reclamation waits for the entry to expire (7 days), so an operator prepays for the nodes they run **plus a spare** — typically two: the live node and its migration target.

### API

| Method | Kind | Purpose |
|---|---|---|
| `prepay_attestation_storage(account_id, grants)` | `#[payable]` | Adds `grants` to `account_id`. Requires an attached deposit of exactly `fee × grants` and rejects anything else, so there is no remainder to keep or refund. Permissionless — anyone may prepay for any account. |
| `available_attestation_grants(account_id) -> u32` | view | Grants available. |

> | `attestation_storage_fee() -> NearToken` | view | Current fee. |

Reviewer: we might be fine skipping this one. If this is anyway a manual operation by operators, they should read the doc in the `prepay_attestation_storage` to know exactly how much to pay anyway

Author: do you mean skip this attestation_storage_fee? I think is better to have this in the contract. (docs can get stale)

Author: Keeping it, but with a better reason than my "docs get stale" — the doc here is *guaranteed* to go stale. The fee is votable, so it changes without a release, and `prepay_attestation_storage` requires exactly `fee × grants` and rejects anything else. A number published in a doc therefore produces failed transactions, not overpayments.

Your point does land in one way: `config()` already exposes the field, so this view isn't the only on-chain source. The doc now says that, and justifies the view as convenience — an operator scripting `attached-deposit` shouldn't have to parse the whole config blob.

Reviewer: >  The fee is votable, so it changes without a release

Right, we are making it votable :) Probably we didn't need that though, I don't see operators ever changing that, but that's fine. But again, if the operator can already get the value from `config()` then we don't need this convenience. Extracting the value from the config blob is trivial, and we do not expect any operator to do this automatically anyway, as this is meant to be a manual step after all, specially for a value that would almost never change.

Author: Nothing changed yet — I flagged this rather than acting on it, since it is permanent public API.

I am fine with removing it. Your argument holds: `config()` already returns the field, the operator reads it once by hand as part of a manual step, and the value will almost never change. That does not pay for a dedicated public method, an ABI entry and a thing to keep working forever.

@netrome — a second opinion before it goes? You were the one who pushed for the fee to be configurable rather than hard-coded, so you may have a view on whether a dedicated read is worth keeping alongside `config()`.

I will start the implementation with the view in place, since dropping it later is a one-line removal plus an ABI snapshot regeneration. If we settle on removing it, that happens before the implementation PR merges rather than after.

Reviewer 2: Yeah if it's in the config, there's no need for a dedicated view function.

1. **Entry exists and this account owns it** — re-attestation. Updates in place, consumes nothing. Keys on *ownership*, not key presence: a presence-only test would classify someone else's key as "no grant needed".
2. **New entry** — consume one grant; reject if the account has none.
3. **Entry reclaimed** by `clean_invalid_attestations` — return one grant to its owner. There is exactly one removal site, so the counter cannot drift. This adds a grants-map write per removed entry — a row *insert*, not an update, whenever the owner's row was deleted at zero — inside a gas-bounded sweep, so `clean_invalid_attestations_tera_gas` and `RESHARE_CLEAN_INVALID_ATTESTATIONS_MAX_SCAN` need re-validating against the heavier per-entry cost.

Consuming at insert means a failed attestation consumes nothing — nothing was stored, so nothing is owed — and keeps [#3991](https://github.com/near/mpc/issues/3991) unblocked, since no charge has to survive a failing callback.

Migration and multi-node operators need no special rule: prepay again. Hence no cap constant, no participant-status check, and no assumption about how many nodes share an account.

### Fee

> 0.02 NEAR, about 2.7× the floor. Figures are **charged** bytes — key and record overhead included — not borsh sizes:

Reviewer 2: Implementation note: I think this should be configurable on the contract and not hard-coded.

Author: Agreed, and that's the design already — it's a governance-votable `Config` field, not a constant. The doc said so eight lines below your comment (and again in the Decisions table), which is too far from the number to count.

Moved it onto the number itself: "0.02 NEAR — a governance-votable `Config` field, not a constant — about 2.7× the floor."

| Component | Charged bytes | Cost |
|---|---|---|
| Worst-case entry: a `Mock` one at 604 (`Dstack` is 599) | 604 | 0.00604 NEAR |
| Grants-map row | ~130 | ~0.0013 NEAR |

> | **Floor** | **~734** | **~0.0073 NEAR** |

Reviewer: this is not exactly true in the sense that when we insert elements in a datastructure, such as a HashMap, more bytes are probably needed. The item's size is just the best possible case. It would be good to backup the values we end up using with some benchmark using many attestations.

Author: this is why I put 0.02 near as the suggested value- it give us a lot of buffer.

Author: Correcting my earlier reply — the buffer isn't the argument, and the numbers aren't best-case. 604 is not a computed borsh size: `measure_stored_entry_bytes` (`crates/contract/src/lib.rs`) inserts into the real `stored_attestations`, calls `.flush()`, and returns the `env::storage_usage()` delta. That's the charged cost with the `IterableMap` record and key overhead included — the benchmark you're asking for, already run, and pinned by a test at 604 mock / 599 dstack whose doc comment forbids updating the number to make a failure pass.

The doc asserted "charged bytes" without saying where they came from. It now cites the measurement. The grants-map row (~130) is genuinely an estimate, and the doc now says so; the headroom covers it.

Reviewer: This benchmark is still not so accurate for what we need. The underlying data structure is a hash map, so its total size "per entry" is not as easy to compute. To really figure this out we would need to insert many attestations and then get the average or maximum of the cost per attestation. Therefore, we still don't have proper numbers, but we could do it as part of the PR following this design. No need to mention an amount here, except as a rough estimate

Reviewer 2: Yeah I think this is also why it's worth keeping this configurable.

The rest is headroom, so a layout change cannot leave sold grants under-funded. Over-sizing costs nothing — the margin is never returned — while under-sizing silently reopens the drain. The fee itself is a governance-votable `Config` field, so it can be re-priced without a release.

### Flow

```mermaid
---
title: Onboarding — operator prepays one grant, node self-submits
---
sequenceDiagram
    box Operator Secure Environment
        participant OP as Operator<br/>(full-access key)
    end
    box MPC Node CVM (TDX)
        participant NODE as MPC Node<br/>(function-call key)
    end
    box NEAR Blockchain
        participant BC as mpc-contract
    end

    OP ->> BC: prepay_attestation_storage(node_account, 1) + fee
    BC ->> BC: grants[node_account] += 1

    NODE ->> BC: submit_participant_info(attestation, tls_pubkey)
    BC ->> BC: precondition: new entry, owned by caller, grant available?<br/>rejects here if not, before any verification
    BC ->> BC: verify quote, report_data == hash(tls_pk, signer_pk)
    BC ->> BC: consume one grant, insert entry
    BC -->> NODE: ok

> ```

Reviewer: it would be more useful to explain that before setting up the node, or starting it, the operator needs to submit the deposit. That is not clear from this graph

Author: in my mind I thought the step would be near the add key to your contract, but actually you are right, this step be done before.

Author: Fixed — the diagram now puts the prepayment before the CVM exists, with notes marking the two phases:

```
Note over OP,BC: Before the CVM is started — the prepayment needs
                 only the account id, not the node's key
OP ->> BC: prepay_attestation_storage(node_account, 1) + fee
...
Note over NODE,BC: After start-up and indexer sync
```

Without the prepayment the submission fails immediately and the node retries in a loop until granted.

## Operator UX

> One new step in the [operator guide](../running-an-mpc-node-in-tdx-external-guide.md), after [Add the Node Account Key](../running-an-mpc-node-in-tdx-external-guide.md#add-the-node-account-key-to-your-account) and before [Submitting Participant Info](../running-an-mpc-node-in-tdx-external-guide.md#submitting-participant-info). Every other step is unchanged, including the off-chain `attestation-cli` check, which this design does not touch.

Reviewer: by the time you can get the node account, that node would already be sending attestations that would fail. So this new step should be done before starting the node.

Author: not exactly, the node account/key - are exposed in /public_data very soon after node start-up but attestation is send only after the node is sync (at this is what I recall). - but as I said in the comment above, this step can be done sooner.

Author: the doc now anchors the new step to node-account creation

```bash
near contract call-function as-transaction \
  v1.signer prepay_attestation_storage \
  json-args '{"account_id":"<your-node-account>","grants":2}' \
  prepaid-gas '30.0 Tgas' attached-deposit '0.04 NEAR' \
  sign-as <your-operator-account> network-config mainnet sign-with-keychain send
```

Ask for as many grants as the operator runs nodes, plus the spare — the example buys two. Existing attested nodes need no action; they re-attest under rule 1.

## Decisions

| Decision | Chosen | Instead of | Why |
|---|---|---|---|
| Accounting | Grant counter | Per-account NEAR balance, NEP-145 shaped | No arithmetic, no amounts stored, and a fee change cannot strand a sold grant. |
| Multi-node / migration | Prepay per node | Free extra entry gated on `ongoing_migrations`; or a hard cap of N | Both needed a participant check or a magic number; repeating the prepayment needs neither. |

> | Grant semantics | Capacity — reclaiming returns it | Consumable ticket | The contract got the storage back, so charging again charges twice. Also lets an operator re-provision on the grants they hold. |

Reviewer: this is a bit confusing to me
```suggestion
| Grant semantics | Capacity — removing/expiring attestation returns it | Consumable ticket | The contract got the storage back, so charging again charges twice. Also lets an operator re-provision on the grants they hold. |
```

Author: Taken, with one correction — "expiring" on its own doesn't return the grant. Expiry only makes the entry sweepable; the grant comes back when `clean_invalid_attestations` actually removes it. That gap is exactly why the design tells operators to buy a spare, so the row shouldn't blur it:

| Grant semantics | Capacity — the grant returns when the entry is swept | Consumable ticket | The contract got the storage back, so charging again charges twice. Expiry alone does not return it, which is why an operator keeps a spare. |

You were right that "reclaiming" was doing unexplained work in that cell.

| Cap entries per account | Implicit accounts are ~free and deletable, so a cap moves amplification only 7.2× → ~5.8×. |
| Owner-callable release returning a grant early | Removes the spare-grant need and reclaims storage sooner, but is another entry point that must refuse to evict a live participant. Revisit if the 7-day wait hurts. |
| Restrict submission to current-or-proposed participants | Not possible today — `vote_new_parameters` rejects proposals naming unattested participants, so attestation must come first. Would need a gate at the Running→Resharing transition. Worth its own issue: it removes the attack surface rather than pricing it. |

## Security

Every new entry consumes a paid grant, so the contract funds no new attestation storage and amplification drops below 1. Recycling does not weaken that — a grant returns only once its entry is gone, so `entries held ≤ grants bought` holds throughout. It does permit unlimited churn on a fixed number of grants, which is harmless because each cycle requires the previous entry to be reclaimed first.

Two accepted residuals:

> - **Pre-upgrade entries yield a grant when swept**, since rule 3 cannot distinguish them from paid ones. Accepted: there are 14 such entries on mainnet and 31 on testnet, so the free capacity is negligible (and init-time mock sentinels never expire, so they are never swept and never yield one). Check the count at deploy though — a count in the thousands would mean the still-open drain was exploited first, and those entries should be purged rather than granted.

Reviewer: why is this an accepted residual? we can just do a contract migration to cover existing ones no? Also we are marking all attestations with an expire date soon, so at that time this is no longer true

Author: Maybe I didn't explain this right,
existing entries -> are just kept, nothing that we need to do. (it is like we just give them the grant for free, and this is totally fine). I explicitly motioned since one of the bot flag it before.
but we should check before the migration that the number of entries matches the expected number of entries. (in case there is some spamming until we release this fix.

Author: Adopted your migration suggestion — it's better than what I defended, and cheaper than the alternative I rejected.

The migration now credits one grant to each account already holding an entry. Rule 3 then returns a grant that was genuinely issued, so the invariant holds for old and new entries alike, with no per-entry "was this paid for?" marker and no second map (the ~110 bytes/entry option I'd turned down on cost). One residual drops off the list.

Two things it doesn't change, so the doc still says them: the free capacity is the same 14 mainnet / 31 testnet either way — the migration just issues it explicitly and auditably up front instead of drip-feeding it at sweep time — and the deploy-time count check stays, since that's what catches a pre-upgrade spam flood.

On the expiry-date point: that already landed, and it doesn't quite close the case. `stamp_expiry_on_legacy_mocks` is on `main` (`crates/contract/src/v3_13_0_state.rs:122`), but `with_mocked_participant_attestations` still inserts non-expiring `MockAttestation::Valid` sentinels at init (`crates/contract/src/tee/tee_state.rs:136-142`), so those are never swept. Grandfathering handles them cleanly: the sentinel's account gets its grant, and a real submission for the same TLS key is rule 1, which consumes nothing.

Reviewer: > but with_mocked_participant_attestations still inserts non-expiring MockAttestation::Valid sentinels at init (crates/contract/src/tee/tee_state.rs:136-142), so those are never swept. Grandfathering handles them cleanly: the sentinel's account gets its grant, and a real submission for the same TLS key is rule 1, which consumes nothing.

oh, if that is true, why didn't we close that gap in https://github.com/near/mpc/pull/3785 ?

Author: Please disregard my previous reply on this thread — it was drafted by an agent and I posted it without reviewing it first.

Two things:

**1. The migration in this PR is deliberately simple.** Existing attestations stay in place and get no special handling: an entry already holds a slot that no grant was ever bought for, so we treat it as already paid for. No crediting step, no per-entry marker, no second map. Re-attestation under the same TLS key is rule 1, which consumes nothing, so those operators need no grant and no action from us.

**2. `with_mocked_participant_attestations` runs only at init, so no real network is affected.** Its only non-test callers are `init` and `init_running`, both `#[init]` without `ignore_state`, so they panic if state already exists — mainnet and testnet are long past that point and nothing re-runs them. The non-expiring sentinels therefore only exist in freshly initialised state: localnet, e2e and the sandbox tests. They are also overwritten as soon as the real node submits for the same TLS key.

That is why I left the init path alone in #3785 — I wanted that PR limited to state that actually exists on a live network. #1087 already tracks the init-time attestation question more broadly.

> - **The contract carries the price risk on sold grants**, bounded by the number outstanding. It stores counts, not purchase prices, so it cannot retro-price them; a future update can address this if the gap becomes material.

Reviewer: I know what this means, but reads really cumbersome. Just saying that if storage prices change in the future we might need to adapt.

Author: Fair — you got it and still had to work for it. Rewritten:

> **If storage gets more expensive, grants already sold may not cover their entry** — the contract absorbs the difference, bounded by the number of grants outstanding. It stores counts, not purchase prices, so sold grants cannot be re-priced; revisit if the gap becomes material.

### API

| Method | Kind | Purpose |
|---|---|---|
| `prepay_attestation_storage(account_id, grants)` | `#[payable]` | Adds `grants` to `account_id`. Requires an attached deposit of exactly `fee × grants` and rejects anything else, so there is no remainder to keep or refund. Permissionless — anyone may prepay for any account. |
| `available_attestation_grants(account_id) -> u32` | view | Grants available. |
| `attestation_storage_fee() -> NearToken` | view | **Optional — open question.** The fee is already returned by `config()`, and reading it is a manual step on a value that almost never changes, so a dedicated method may not earn its permanent place in the API. Left out of the implementation until decided; adding it later is cheaper than removing it. |

An operator reads the current fee from `config()` and attaches the exact multiple — the fee is votable, so a number published in a doc instead would eventually produce failed transactions, not overpayments. "No NEAR is ever returned" therefore means simply: no withdrawal method.

### Charging rules

> Evaluated read-only at the top of `submit_participant_info` — before any verification, so an ungranted or unauthorised call never reaches the `Mock` checks or a `verify_quote` round trip — and applied at insert.

Reviewer 2: Nit: Don't see the point in distinguishing the different verifications we have here

```suggestion
Evaluated read-only at the top of `submit_participant_info` — before any verification, so an ungranted or unauthorised call never reaches attestation verification checks.
```

    BC ->> BC: verify quote, report_data == hash(tls_pk, signer_pk)
    BC ->> BC: consume one grant, insert entry
    BC -->> NODE: ok
```

Without the prepayment the submission fails immediately and the node retries in a loop until granted. Following the guide's order avoids that state entirely.

## Existing nodes

Already-attested nodes need no grant and their operators need to do nothing. Their entry is already stored, and re-attesting under the same key updates it in place for free (rule 1) — so an existing entry is a grant already spent: it holds a slot, and none was ever bought for it.

Nothing in the contract treats these entries specially: no migration step, no per-entry marker, no second map. That is the point — for existing nodes this change needs no code at all.

Accepted deliberately: if such an entry is later swept, rule 3 hands its owner a grant they never bought. Negligible at 14 entries on mainnet and 31 on testnet, but check the count before deploying — thousands would mean the drain was exploited first, and those entries should be purged rather than left to become grants.

> ## Operator UX

Reviewer 2: This feels a bit strange but sure, it's worth keeping as a reminder to update the operator guide - although we should always remember to do this for operator-facing changes.

## Design

### Grants, not balances

A grant is a count, not an amount. `prepay_attestation_storage` increments a per-account counter, storing a new entry decrements it, and reclaiming an entry increments it back. The contract never stores NEAR amounts per operator, so there is no arithmetic to get wrong and nothing to reconcile.

A grant is therefore **capacity to hold one entry**, not a one-time ticket. The invariant is:

```text
entries an account currently holds  ≤  grants that account has bought
```

Recycling the grant on reclamation is what makes that work, and it earns its keep twice over:

- **Fairness** — the fee bought storage. Once the contract has that storage back, charging again for the next entry would be charging twice for the same thing.
- **Operator experience** — an operator prepays for the capacity they actually run (three grants for three nodes, say) and is then done with it. Nodes can be torn down, re-provisioned, rotated and migrated indefinitely without another prepayment, as long as no more than three entries are held at once. Without recycling, every re-provisioned node is a fresh purchase and a fresh thing to remember.

> **No NEAR is ever returned.** Recycling gives back capacity, never money: a deposit is consumed permanently the moment it is made, and the only thing that ever comes back is the right to store another entry. There is no withdrawal path and none is planned.

Author: > Consider tightening the wording to distinguish "no withdrawals / redeem" from "refund any overpayment in the prepay call", so the document is internally consistent and aligned with #3972.

Done — the two were conflated. The doc now states that "No NEAR is ever returned" covers both cases: no withdrawal method, and no refund of an overpayment. It also says why keeping the remainder deviates from the contract's usual `require_deposit` + `refund_to` pattern — a grant is a discrete unit, so the leftover is at most one fee short of the next grant, and a transfer path for sub-0.02 NEAR dust is not worth it.

On the #3972 alignment: the issue was the stale side, not the doc. Its first acceptance criterion still required refunding excess, which the design meeting overrode — that AC is now updated, along with three others that had drifted from the agreed model.

> **No NEAR is ever returned.** Recycling gives back capacity, never money: a deposit is consumed permanently the moment it is made, and the only thing that ever comes back is the right to store another entry. There is no withdrawal path and none is planned.

Author: Correction to my reply above: the design changed rather than the wording.

`prepay_attestation_storage` now takes an explicit grant count and requires an attached deposit of exactly `fee × grants`, rejecting anything else:

```
prepay_attestation_storage(account_id, grants)   #[payable]
```

So there is no remainder to keep, and nothing to reconcile against the `require_deposit` + `refund_to` convention — the case your finding was about no longer exists. My earlier reply justified keeping the remainder; disregard that reasoning. "No NEAR is ever returned" now means only that there is no withdrawal method.
